pkg/kubelet/images/pullmanager/locks.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pullmanager
+
+import "sync"
+
+// namedLockSet stores named locks in order to allow to partition context access
+// such that callers that are mutually exclusive based on a string value can
+// access the same context at the same time, compared to a global lock that
+// would create an unnecessary bottleneck.
+type namedLockSet struct {
+	globalLock sync.Mutex
+	locks      map[string]*sync.Mutex
+}
+
+func NewNamedLockSet() *namedLockSet {
+	return &namedLockSet{
+		globalLock: sync.Mutex{},
+		locks:      map[string]*sync.Mutex{},
+	}
+}
+
+func (n *namedLockSet) Lock(name string) {
+	func() {
+		n.globalLock.Lock()
+		defer n.globalLock.Unlock()
+		if _, ok := n.locks[name]; !ok {
+			n.locks[name] = &sync.Mutex{}
+		}
+	}()
+	// This call cannot be guarded by the global lock as it would block the access
+	// to the other locks
+	n.locks[name].Lock()
+}
+
+// Unlock unlocks the named lock. Can only be called after a previous Lock() call
+// for the same named lock.
+func (n *namedLockSet) Unlock(name string) {
+	// cannot be locked by the global lock as it would deadlock once GlobalLock() gets activated
+	if _, ok := n.locks[name]; ok {
+		n.locks[name].Unlock()
```

Comment:
this leaks lock entries from the map... need to delete them somehow

however, that makes concurrent calls to Unlock("foo") and Lock("foo") racy... if the Unlock is going to delete the entry, and Lock is counting on the entry being present

Reply:
We could use lock striping - preallocate, say, 151 locks and then
```go
func (s *stripedLockSet) Lock(name string) {
    fooHash := hashKey(name) // hashKey(string) ->uint32
    s.locks[fooHash % s.locksNum].Lock()
}

func (s *stripedLockSet) GlobalLock() {
    for i := range s.locks {
        s.locks[i].Lock()
    }
}
```
We'd no longer need an additional global lock as we don't need to prevent anyone from adding new locks into the struct and it's no longer unbounded.